test-data/unit/check-generics.test

```diff
@@ -1815,3 +1815,214 @@ def g(x: T) -> T: return x
+reveal_type(B.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.int*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.int*'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOfGenericClassMultipleInheritance]
+from typing import TypeVar, Generic
+T = TypeVar('T')
+class A(Generic[T]):
+    @classmethod
+    def f(cls) -> T: pass
+class B(Generic[T]):
+    @classmethod
+    def g(cls) -> T: pass
+class C(A[str], B[int]):
+    @classmethod
+    def f(cls) -> str: pass
+    @classmethod
+    def g(cls) -> int: pass
+reveal_type(C.f())  # E: Revealed type is 'builtins.str'
+reveal_type(C.g())  # E: Revealed type is 'builtins.int'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOfSubcClassOfSubClassOfGenericClass]
+from typing import TypeVar, Generic
+T = TypeVar('T')
+class A(Generic[T]):
+    @classmethod
+    def f(cls) -> T: ...
+class B(A[str]): pass
+class C(B): pass
+class D(C): pass
+reveal_type(B.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(B.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(D.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(D.f())  # E: Revealed type is 'builtins.str*'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOfGenericClassMultipleLevelInheritance]
+from typing import TypeVar, Generic
+T = TypeVar('T')
+class A(Generic[T]):
+    @classmethod
+    def f(cls) -> T: ...
+class B(A[str]): pass
+class C(B): pass
+class D(C): pass
+class E(D): pass
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(D.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(D.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(E.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(E.f())  # E: Revealed type is 'builtins.str*'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOfGenericClassMultipleLevelInheritanceWithChangingTypeVariable]
+from typing import TypeVar, Generic
+T = TypeVar('T')
+S = TypeVar('S')
+class A(Generic[T]):
+    @classmethod
+    def f(cls) -> T: ...
+class B(A[S]): pass
+class C(B[str]): pass
+class D(B[int]): pass
+reveal_type(B[str].f())  # E: Revealed type is 'builtins.str'
+reveal_type(B[int].f())  # E: Revealed type is 'builtins.int'
+reveal_type(C.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(D.f())  # E: Revealed type is 'builtins.int*'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOfMultipleGenericClass]
+from typing import TypeVar, Generic
+T = TypeVar('T')
+class A(Generic[T]):
+    @classmethod
+    def f(cls) -> T: ...
+class B(Generic[T]):
+    @classmethod
+    def f(cls) -> T: ...
+class C(A[str], B[int]): pass
+class D(A[int], B[str]): pass
```

Comment:
I actually wanted to see a test where a class method return type depends on type variables from two different bases.

Reply:
@ilevkivskyi

Does you mean like this?

```python
from typing import Tuple, TypeVar, Generic
T = TypeVar('T')
S = TypeVar('S')


class A(Generic[T]):
    @classmethod
    def f(cls) -> T: ...

class B(Generic[T]):
    @classmethod
    def g(cls) -> T: ...

class C(A[T], B[S]):
    @classmethod
    def h(cls) -> Tuple[T, S]: ...

reveal_type(C[str, int].h())  # E: Revealed type is 'Tuple[builtins.str, builtins.int]'
reveal_type(C[int, str].h())  # E: Revealed type is 'Tuple[builtins.int, builtins.str]'
```

Reply:
Yes. But I would add one more inheritance:
```python
class D(C[int, str]):
    ...
reveal_type(D.h())
```
I think accessing it on directly on `C[str, int]` should be an error.

Reply:
@ilevkivskyi 

> I think accessing it on directly on `C[str, int]` should be an error.

Sorry I don't understand it. Why should `C[str, int]` be an error?
I think it seems to be a valid declaration.

Reply:
Not `C[str, int]` itself, expression like `C[str, int].h()` should be an error (probably saying something about ambiguous access, etc.)

Reply:
@ilevkivskyi 

Thank you. But I still have not understood that it is ambiguous.
I think this case is not ambiguous as same as following case is valid.

```python
from typing import TypeVar, Generic

T = TypeVar('T')


class A(Generic[T]):
    @classmethod
    def f(cls) -> T: ...


reveal_type(A[str].f())  # E: Revealed type is 'builtins.str'
```

Is this behavior correct?

Reply:
No, both should be errors. Only subclassing can actually bind a type variable. At runtime, `A[str]` is essentially just an alias to `A` there is no way accessing something on `A[str]` can be different from accessing it on `A`.

On the contrary, this _is_ valid, because there is a new class that (potentially) can add something:
```python
class B(A[str]):
    ...
B.f()  # OK
```
we still can't guarantee it is safe, but I think we should not flag this.

Reply:
Oh, thank you. I understand.
And I found that I must fix other test cases that I've added too.